File: 0086-partition-list/0086-partition-list.java
class Solution {
    public ListNode partition(ListNode head, int x) {
        // Dummy nodes to start the before and after lists
        ListNode beforeHead = new ListNode(0);
        ListNode afterHead = new ListNode(0);

        // Pointers to the current end of the before and after lists
        ListNode before = beforeHead;
        ListNode after = afterHead;

        // Traverse the original list
        while (head != null) {
            if (head.val < x) {
                before.next = head;
                before = before.next;
            } else {
                after.next = head;
                after = after.next;
            }
            head = head.next;
        }

        // Important: Terminate the after list
        after.next = null;

        // Connect the two partitions
        before.next = afterHead.next;

        // The start of the new list is at beforeHead.next
        return beforeHead.next;
    }
}

File: 0086-partition-list/0086-partition-list.c
struct ListNode* partition(struct ListNode* head, int x) {

    if (!head) return NULL;

    struct ListNode *dummySmall = malloc(sizeof(struct ListNode));
    struct ListNode *dummyLarge = malloc(sizeof(struct ListNode));
    dummySmall->next = NULL;
    dummyLarge->next = NULL;

    struct ListNode *small = dummySmall;
    struct ListNode *large = dummyLarge;
    struct ListNode *temp  = head;

    while (temp) {
        if (temp->val < x) {
            small->next = temp;
            small = small->next;
        } else {
            large->next = temp;
            large = large->next;
        }
        temp = temp->next;
    }

    large->next = NULL;            // end the list
    small->next = dummyLarge->next;

    struct ListNode* ans = dummySmall->next;

    free(dummySmall);
    free(dummyLarge);
    return ans;
}
